Mathlib/CategoryTheory/FiberedCategory/Cartesian.lean

```diff
@@ -115,6 +144,197 @@ instance of_comp_iso {b' : 𝒳} (φ' : b ≅ b') [IsHomLift p (𝟙 S) φ'.hom]
+      apply map_uniq
+      apply map_uniq
+      simp only [assoc, hπ'₂]
+
+/-- Given two commutative squares
+```
+a --φ--> b --ψ--> c
+|        |        |
+v        v        v
+R --f--> S --g--> T
+```
+such that `φ ≫ ψ` and `ψ` are strongly cartesian, then so is `φ`. -/
+protected lemma of_comp [IsStronglyCartesian p g ψ] [IsStronglyCartesian p (f ≫ g) (φ ≫ ψ)]
+    [IsHomLift p f φ] : IsStronglyCartesian p f φ where
+  universal_property' := by
+    intro a' h τ hτ
+    have h₁ : IsHomLift p (h ≫ f ≫ g) (τ ≫ ψ) := by simpa using IsHomLift.comp p (h ≫ f) _ τ ψ
+    /- We get a morphism `π : a' ⟶ a` such that `π ≫ φ ≫ ψ = τ ≫ ψ` from the universal property
+    of `φ ≫ ψ`. This will be the morphism induced by `φ`. -/
+    use map p (f ≫ g) (φ ≫ ψ) (f' := h ≫ f ≫ g) rfl (τ ≫ ψ)
+    refine ⟨⟨inferInstance, ?_⟩, ?_⟩
+    /- The fact that `π ≫ φ = τ` follows from `π ≫ φ ≫ ψ = τ ≫ ψ` and the universal property of
+    `ψ`. -/
+    · apply IsStronglyCartesian.ext p g ψ (h ≫ f) (by simp)
+    -- Finally, the uniqueness of `π` comes from the universal property of `φ ≫ ψ`.
+    · intro π' ⟨hπ'₁, hπ'₂⟩
+      apply map_uniq
+      simp [hπ'₂.symm]
+
+end
+
+section
+
+variable {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S)
+
+instance of_iso (φ : a ≅ b) [IsHomLift p f φ.hom] : IsStronglyCartesian p f φ.hom where
+  universal_property' := by
+    intro a' g τ hτ
+    use τ ≫ φ.inv
+    refine ⟨?_, by aesop_cat⟩
+    simpa using (IsHomLift.comp p (g ≫ f) (isoOfIsoLift p f φ).inv τ φ.inv)
+
+instance of_isIso (φ : a ⟶ b) [IsHomLift p f φ] [IsIso φ] : IsStronglyCartesian p f φ :=
+  @IsStronglyCartesian.of_iso _ _ _ _ p _ _ _ _ f (asIso φ) (by aesop)
+
+/-- A strongly cartesian morphism lying over an isomorphism is an isomorphism. -/
+lemma isIso_of_base_isIso (φ : a ⟶ b) [IsStronglyCartesian p f φ] [IsIso f] : IsIso φ := by
+  subst_hom_lift p f φ; clear a b R S
+  -- Let `φ` be the morphism induced by applying universal property to `𝟙 b` lying over `f⁻¹ ≫ f`.
+  let φ' := map p (p.map φ) φ (IsIso.inv_hom_id (p.map φ)).symm (𝟙 b)
+  use φ'
+  -- `φ' ≫ φ = 𝟙 b` follows immediately from the universal property.
+  have inv_hom : φ' ≫ φ = 𝟙 b := fac p (p.map φ) φ _ (𝟙 b)
+  refine ⟨?_, inv_hom⟩
+  -- We will now show that `φ ≫ φ' = 𝟙 a` by showing that `(φ ≫ φ') ≫ φ = 𝟙 a ≫ φ`.
+  have h₁ : IsHomLift p (𝟙 (p.obj a)) (φ  ≫ φ') := by
+    rw [← IsIso.hom_inv_id (p.map φ)]
+    apply IsHomLift.comp
+  apply IsStronglyCartesian.ext p (p.map φ) φ (𝟙 (p.obj a))
+  simp only [assoc, inv_hom, comp_id, id_comp]
+
+end
+
+/-- The canonical isomorphism between the domains of two strongly cartesian morphisms lying over
+isomorphic objects. -/
+noncomputable def domainIsoOfBaseIso {R R' S : 𝒮} {a a' b : 𝒳} {f : R ⟶ S} {f' : R' ⟶ S}
+  {g : R' ≅ R} (h : f' = g.hom ≫ f) (φ : a ⟶ b) (φ' : a' ⟶ b) [IsStronglyCartesian p f φ]
+    [IsStronglyCartesian p f' φ'] : a' ≅ a where
+  hom := map p f φ h φ'
+  inv := @map _ _ _ _ p _ _ _ _ f' φ' _ _ _ _ _ (congrArg (g.inv ≫ ·) h.symm) φ
```

Comment:
Is there a way to avoid the `@map` syntax here?

Reply:
Yes, fixed!